### PR TITLE
Implementaiton of selected() optimizations on android for index queries

### PR DESCRIPTION
**Pull request by ctsims**
cross-request: https://github.com/dimagi/commcare-core/pull/508

### CODE REVIEW

File: app/src/org/commcare/activities/components/FormEntryInstanceState.java [hunk 1]
```diff
@@ -195,6 +196,6 @@ public void setFormPath(String path) {
     }
 
     public static String getInstanceFolder() {
-        return mInstancePath.substring(0, mInstancePath.lastIndexOf("/") + 1);
+        return mInstancePath.substring(0, mInstancePath.lastIndexOf(File.separator) + 1);
```
Comment (ctsims) [inline app/src/org/commcare/activities/components/FormEntryInstanceState.java]: these changes just fix unit tests for windows

File: app/src/org/commcare/engine/cases/AndroidCaseInstanceTreeElement.java [hunk 2]
```diff
@@ -91,8 +95,9 @@ protected synchronized void loadElements() {
         String[] namesToMatch = new String[numKeys];
         String[] valuesToMatch = new String[numKeys];
         for (int i = numKeys - 1; i >= 0; i--) {
-            namesToMatch[i] = keys.elementAt(i);
-            valuesToMatch[i] = (String)values.elementAt(i);
+            namesToMatch[i] = profiles.elementAt(i).getKey();
+            valuesToMatch[i] = (String)
```
Comment (wpride) [inline app/src/org/commcare/engine/cases/AndroidCaseInstanceTreeElement.java]: Why do we know this to be a String? OR why does IndexedValueLookup allow the ambiguity of an Object?

File: app/src/org/commcare/engine/cases/AndroidCaseInstanceTreeElement.java [hunk 3]
```diff
@@ -102,53 +107,69 @@ protected synchronized void loadElements() {
...
+                //remove the match from the inputs
+                optimizations.removeElementAt(0);;
+                return mIndexCache.get(indexCacheKey);
+            }
+
+            matchingCases = mCaseIndexTable.getCasesMatchingIndex(indexName, value);
+        }
+        if(op instanceof IndexedSetMemberLookup) {
+            IndexedSetMemberLookup sop = (IndexedSetMemberLookup)op;
+            matchingCases = mCaseIndexTable.getCasesMatchingValueSet(indexName,sop.valueSet);
```
Comment (wpride) [inline app/src/org/commcare/engine/cases/AndroidCaseInstanceTreeElement.java]: missing space after indexName,

File: app/src/org/commcare/engine/cases/AndroidCaseInstanceTreeElement.java [hunk 4]
```diff
@@ -102,53 +107,69 @@ protected synchronized void loadElements() {
...
+            //Check whether we've got a cache of this index.
+            if (mIndexCache.containsKey(indexCacheKey)) {
+                //remove the match from the inputs
+                optimizations.removeElementAt(0);;
+                return mIndexCache.get(indexCacheKey);
+            }
+
+            matchingCases = mCaseIndexTable.getCasesMatchingIndex(indexName, value);
+        }
+        if(op instanceof IndexedSetMemberLookup) {
```
Comment (wpride) [inline app/src/org/commcare/engine/cases/AndroidCaseInstanceTreeElement.java]: missing whitespace

File: app/src/org/commcare/models/database/user/models/CaseIndexTable.java [hunk 5]
```diff
@@ -112,4 +117,38 @@ public void clearCaseIndices(int recordId) {
...
+        args[0] = indexName;
+        for(int i =0 ; i < targetValueSet.length ; ++i) {
+            args[i + 1] = targetValueSet[i];
+        }
+        String inSet = getInSet(targetValueSet.length);
+        if (SqlStorage.STORAGE_OUTPUT_DEBUG) {
+            String query = String.format("SELECT %s FROM %s WHERE %s = ? AND %s in " + inSet, COL_CASE_RECORD_ID, TABLE_NAME, COL_INDEX_NAME, COL_INDEX_TARGET);
+            DbUtil.explainSql(db, query, args);
+        }
+        Cursor c = db.query(TABLE_NAME, new String[]{COL_CASE_RECORD_ID}, COL_INDEX_NAME + " = ? AND " + COL_INDEX_TARGET + " in " + inSet, args, null, null, null);
```
Comment (wpride) [inline app/src/org/commcare/models/database/user/models/CaseIndexTable.java]: String formatting as above would be nice

File: app/src/org/commcare/models/database/user/models/CaseIndexTable.java [hunk 6]
```diff
@@ -112,4 +117,38 @@ public void clearCaseIndices(int recordId) {
...
+        return SqlStorage.fillIdWindow(c, COL_CASE_RECORD_ID);
+    }
+
+    public static String getInSet(int number) {
```
Comment (wpride) [inline app/src/org/commcare/models/database/user/models/CaseIndexTable.java]: Nice function, small DRY refactor for future PR:
 https://github.com/dimagi/commcare-core/blob/a2dbf99080e5121e674dc5c809a4f4ea7383954c/src/api/java/org/commcare/api/persistence/SqlHelper.java#L144-L144
https://github.com/dimagi/commcare-core/blob/cd7fc0ef293d1d58c00a8320661142557c6d0258/src/main/java/org/commcare/modern/database/TableBuilder.java#L164-L164

File: app/src/org/commcare/engine/cases/AndroidCaseInstanceTreeElement.java (hunk 2 above)
Comment (ctsims) [inline app/src/org/commcare/engine/cases/AndroidCaseInstanceTreeElement.java]: Hm, this is a good question, this is pretty sketchy, actually, although I'm a bit hesitant to change the old behavior right now. I'm going to add this to a round 2 backlog for revisions that I'm going to make to make all of this indexing stuff less crazy.

**Comment by ctsims** [comment]
@wpride Thanks! Made changes based on your feedback

**Comment by ctsims** [comment]
@damagatchi retest this please

**Comment by ctsims** [comment]
damnit...

**Pull request merged**